mem/mem.go

```diff
@@ -0,0 +1,325 @@
+// pass it through to the places it is needed, however this will prove difficult for
+// existing applications wishing to migrate to the Pool API.
+//
+// The Global Pool is disabled by default. To enable pooling in your application using
+// the Get* generated functions for each message, call Global().Enable() at initialization.
+func Global() *Pool {
+	return globalPool
+}
+
+// Bytes represents a byte slice created from a Pool.
+//
+// Only create these from Pools.
+type Bytes struct {
+	segListPool *segListPool
+	value       []byte
+	valueLen    int
+}
+
+// NewUnmanagedBytes creates a new Bytes that is not backed by any Pool.
+//
+// Users should generally not call this directly.
+func NewUnmanagedBytes(valueLen int) *Bytes {
+	return &Bytes{
+		value:    make([]byte, valueLen),
+		valueLen: valueLen,
+	}
+}
+
+// Value gets the value.
+func (b *Bytes) Value() []byte {
+	return b.value[:b.valueLen]
+}
+
+// Len gets the length.
+//
+// This is equivalent to len(b.Value()).
+func (b *Bytes) Len() int {
+	return b.valueLen
+}
+
+// Truncate truncates the value to the given length.
+func (b *Bytes) Truncate(valueLen int) {
+	b.valueLen = valueLen
+}
+
+// MemsetZero sets the data inside the value to zero.
+//
+// Optimized per https://golang.org/cl/137880043
+func (b *Bytes) MemsetZero() {
+	for i := range b.value {
+		b.value[i] = 0
+	}
+}
+
+// Recycle recycles the Bytes.
+//
+// If this is nil, this is a no-op.
+func (b *Bytes) Recycle() {
+	if b == nil {
+		return
+	}
+	// note this is possible if we did not have a big enough seg list
+	if b.segListPool != nil {
+		b.segListPool.put(b)
+	}
+}
+
+// PooledMessage is a message that can be pooled.
+//
+// This generally is meant for proto.Messages but is compatible with
+// anything that has a Reset and Recycle method.
+type PooledMessage interface {
+	Reset()
+	Recycle()
+}
+
+// Pool is a pool of PooledMessages and Bytes.
+type Pool struct {
+	enabled            uint32
+	channelSize        uint16
+	sortedSegListSizes []int
+
+	messageTypeToMessagePool map[string]*messagePool
+	segListSizeToSegListPool map[int]*segListPool
+
+	lock *sync.RWMutex
+}
+
+// PoolOption is an option when constructing a new Pool.
+type PoolOption func(*Pool)
+
+// WithChannelSize returns a PoolOption that sets the Pool channel size.
+//
+// If your workflow usually only has one object of a given type allocated,
+// the garbage collector will generally garbage collect the object if it
+// is in a sync.Pool before it can be allocated again from the Pool. If this
+// option is used, a channel of allocated PooledMessages of this buffered
+// size will be used to recycle PooledMessages before using the sync.Pool.
+//
+//
+// By default, use DefaultChannelSize. Set this to 0 to not use any channel.
+func WithChannelSize(channelSize uint16) PoolOption {
+	return func(pool *Pool) {
+		pool.channelSize = channelSize
+	}
+}
+
+// WithSegListSizes returns a PoolOption that sets the seg list sizes
+// for allocating Bytes.
```

Comment:
I don't know what `seg list sizes` means. Could you elaborate a little in this comment? What compromises would the user make by changing this?